aten/src/ATen/native/cuda/MultiTensorApply.cuh

```diff
@@ -38,13 +49,213 @@ __device__ __forceinline__ void load_store(
+// buffer. When the combined size of the vectors is small enough, they are
+// packed into the struct itself, which is passed to the kernel through the
+// static kernel argument space. Otherwise, the arguments are packed into a
+// dynamically allocated buffer.
+//
+// NOTE: Previously, we divided a problem into multiple kernel launches so that
+// the arguments for every launch would fit in the static kernel argument
+// space. This approach avoided the need for dynamically allocated kernel
+// arguments, but it led to multiple kernel launches and lower sustained
+// occupancy.
+struct DevArrayPack {
+  static constexpr size_t max_arrays = 8;
+  // The buffer size is selected to ensure that the combined argument size of
+  // multi_tensor_apply_dev_array_pack_kernel does not exceed
+  // max_kernel_arg_size for all template specializations. This is enforced at
+  // compile time with a static assertion.
+  static constexpr size_t small_buffer_size = 3360;
+
+  char small_buffer[small_buffer_size];
+  // i-th array => (buffer_ptr ? buffer_ptr : small_buffer) + offsets[i]
+  size_t offsets[max_arrays];
+  // When small_buffer is used, buffer_ptr is nullptr
+  char* buffer_ptr = nullptr;
+
+  template <typename T>
+  __device__ __forceinline__ void get_array(T*& out, int idx) {
+    if (buffer_ptr == nullptr) {
+      out = reinterpret_cast<T*>(small_buffer + offsets[idx]);
+    } else {
+      out = reinterpret_cast<T*>(buffer_ptr + offsets[idx]);
+    }
+  }
+};
+
+template <typename T>
+void pack_vectors_helper(
+    std::vector<const void*>& ptrs,
+    std::vector<size_t>& sizes,
+    std::vector<size_t>& offsets,
+    size_t& total_bytes,
+    const std::vector<T>& vec) {
+  ptrs.push_back(vec.data());
+  // Align the offset by sizeof(T)
+  total_bytes = at::round_up(total_bytes, sizeof(T));
+  offsets.push_back(total_bytes);
+  sizes.push_back(sizeof(T) * vec.size());
+  total_bytes += sizeof(T) * vec.size();
+}
+
+// Pack multiple vectors into a DevArrayPack
+//
+// When the total size of the input vectors is large, the function packs the
+// vectors into a device buffer allocated with CUDACachingAllocator. Thanks to
+// the allocator's stream awareness, the buffer can be safely accessed by the
+// kernel as long as the tensor owning the buffer is alive at the time of the
+// kernel launch. This tensor is returned to the caller, who is responsible for
+// keeping it alive until the kernel is launched.
+template <int n, typename... Vectors>
+std::tuple<DevArrayPack, c10::optional<at::Tensor>> pack_vectors(
+    const at::Device& device,
+    std::vector<const void*> (&addresses)[n],
+    const Vectors&... vecs) {
+  std::vector<const void*> ptrs;
+  std::vector<size_t> sizes;
+  std::vector<size_t> offsets;
+  size_t total_bytes = 0;
+
+  for (auto d = 0; d < n; ++d) {
+    pack_vectors_helper(ptrs, sizes, offsets, total_bytes, addresses[d]);
+  }
+  (pack_vectors_helper(ptrs, sizes, offsets, total_bytes, vecs), ...);
```

Comment:
What is the `, ...)` for here?

Reply:
This is a fold expression. It calls `pack_vectors_helper` on every element in the parameter pack `vecs`. The expression allows `pack_vectors_helper` to be specialized based on the type of each element in the parameter pack. Without this, we would have to type-erase the elements in `vecs` to make this function reusable for other MTA flavors.

Reply:
if that C++ 17 feature is supported by all the compilers we have in CI (MSVC in particular), I'm happy with it :D

Reply:
Ah good point. Probably not. Will probably need to replace this with some template-based expansion for now.

Reply:
Thanks for raising the point about compiler compatibility. I've replaced this with a template-based expansion